File: backend/src/models/Challan.ts
import { Schema, model, Document, Types } from 'mongoose';

export type PaymentMethod = 'cash' | 'bank_transfer' | 'jazzcash' | 'easypaisa' | 'cheque';
export type ChallanStatus = 'unpaid' | 'partial' | 'paid' | 'waived' | 'overdue';

export interface IChallanItem {
  name: string;
  amount: number;
}

export interface IChallan extends Document {
  orgId: Types.ObjectId;
  branchId: Types.ObjectId;
  studentId: Types.ObjectId;
  classId: Types.ObjectId;
  feeStructureId: Types.ObjectId;
  month: string; // "2025-01" format
  challanNo: string;
  items: IChallanItem[];
  totalAmount: number;
  discount: number;
  waiver: number;
  netAmount: number;
  paidAmount: number;
  dueDate: Date;
  status: ChallanStatus;
  payments: {
    amount: number;
    method: PaymentMethod;
    transactionRef?: string;
    collectedById: Types.ObjectId;
    paidAt: Date;
    receiptNo?: string;
  }[];
  bankName?: string;
  createdAt: Date;
  updatedAt: Date;
}

const challanSchema = new Schema<IChallan>(
  {
    orgId: { type: Schema.Types.ObjectId, ref: 'Organization', required: true },
    branchId: { type: Schema.Types.ObjectId, ref: 'Branch', required: true },
    studentId: { type: Schema.Types.ObjectId, ref: 'Student', required: true },
    classId: { type: Schema.Types.ObjectId, ref: 'Class', required: true },
    feeStructureId: { type: Schema.Types.ObjectId, ref: 'FeeStructure', required: true },
    month: { type: String, required: true },
    challanNo: { type: String, required: true },
    items: [
      {
        name: { type: String, required: true },
        amount: { type: Number, required: true },
      },
    ],
    totalAmount: { type: Number, required: true },
    discount: { type: Number, default: 0 },
    waiver: { type: Number, default: 0 },
    netAmount: { type: Number, required: true },
    paidAmount: { type: Number, default: 0 },
    dueDate: { type: Date, required: true },
    status: {
      type: String,
      enum: ['unpaid', 'partial', 'paid', 'waived', 'overdue'],
      default: 'unpaid',
    },
    payments: [
      {
        amount: { type: Number, required: true },
        method: { type: String, enum: ['cash', 'bank_transfer', 'jazzcash', 'easypaisa', 'cheque'], required: true },
        transactionRef: String,
        collectedById: { type: Schema.Types.ObjectId, ref: 'User', required: true },
        paidAt: { type: Date, default: Date.now },
        receiptNo: String,
      },
    ],
    bankName: String,
  },
  { timestamps: true }
);

challanSchema.index({ orgId: 1, branchId: 1, studentId: 1, month: 1 });
challanSchema.index({ orgId: 1, branchId: 1, month: 1, status: 1 });
challanSchema.index({ orgId: 1, challanNo: 1 }, { unique: true });

export const Challan = model<IChallan>('Challan', challanSchema);

File: mobile/lib/models/challan.dart
class ChallanItem {
  final String name;
  final double amount;
  const ChallanItem({required this.name, required this.amount});
  factory ChallanItem.fromJson(Map<String, dynamic> j) =>
      ChallanItem(name: j['name'] as String, amount: (j['amount'] as num).toDouble());
}

class Challan {
  final String id;
  final String month;
  final String challanNo;
  final List<ChallanItem> items;
  final double totalAmount;
  final double netAmount;
  final double paidAmount;
  final DateTime dueDate;
  final String status; // unpaid | partial | paid | waived | overdue

  const Challan({
    required this.id,
    required this.month,
    required this.challanNo,
    required this.items,
    required this.totalAmount,
    required this.netAmount,
    required this.paidAmount,
    required this.dueDate,
    required this.status,
  });

  double get balance => netAmount - paidAmount;
  bool get isPaid    => status == 'paid' || status == 'waived';
  bool get isOverdue => status == 'overdue';

  factory Challan.fromJson(Map<String, dynamic> j) => Challan(
        id:          j['_id'] as String,
        month:       j['month'] as String,
        challanNo:   j['challanNo'] as String,
        items:       (j['items'] as List? ?? []).map((i) => ChallanItem.fromJson(i as Map<String, dynamic>)).toList(),
        totalAmount: (j['totalAmount'] as num).toDouble(),
        netAmount:   (j['netAmount'] as num).toDouble(),
        paidAmount:  (j['paidAmount'] as num? ?? 0).toDouble(),
        dueDate:     DateTime.parse(j['dueDate'] as String),
        status:      j['status'] as String,
      );
}
